audtorch/datasets/speech_commands.py

```diff
@@ -0,0 +1,117 @@
+import os
+import random
+
+from .utils import (download_url, extract_archive, safe_path)
+from .base import AudioDataset
+
+__doctest_skip__ = ['*']
+
+
+class SpeechCommands(AudioDataset):
+    r"""Data set of spoken words designed for keyword spotting tasks.
+
+    Speech Commands V2 publicly available from Google:
+    http://download.tensorflow.org/data/speech_commands_v0.02.tar.gz
+
+    License: CC BY 4.0
+
+    Args:
+        root (str): root directory of data set, where the CSV files are
+            located, e.g. `/data/speech_commands_v0.02`
+        train (bool, optional): Partition the dataset into the training set.
+            `False` returns the test split. Default: False.
+        download (bool, optional): Download the dataset to `root` if it's not
+            already available. Default: False
+        include (list of str, optional): list of categories to include as
+            commands. If `None` all categories are included. Default:
+            `['yes','no','up','down','left','right','on','off','stop','go']`
+        silence (bool, optional): include a 'silence' class composed of
+            background noise. (Note: use randomcrop when training)
+            Default: `True`
+        transform (callable, optional): function/transform applied on the
+            signal. Default: `None`
+        target_transform (callable, optional): function/transform applied on
+            the target. Default: `None`
+
+    Example:
+        >>> import sounddevice as sd
+        >>> data = SpeechCommands(root='/data/speech_commands_v0.02')
+        >>> print(data)
+        Dataset SpeechCommands
+            Number of data points: 97524
+            Root Location: /data/speech_commands_v0.02
+            Sampling Rate: 16000Hz
+        >>> signal, target = data[4]
+        >>> target
+        'right'
+        >>> sd.play(signal.transpose(), data.sampling_rate)
+    """
+
+    url = ('http://download.tensorflow.org/'
+           'data/speech_commands_v0.02.tar.gz')
+
+    # Available target commands
+    commands = [
+        'right', 'eight', 'cat', 'tree', 'backward',
+        'learn', 'bed', 'happy', 'go', 'dog', 'no',
+        'wow', 'follow', 'nine', 'left', 'stop', 'three',
+        'sheila', 'one', 'bird', 'zero', 'seven', 'up',
+        'visual', 'marvin', 'two', 'house', 'down', 'six',
+        'yes', 'on', 'five', 'forward', 'off', 'four']
+
+    background_noise = '_background_noise_'
+
+    def __init__(self, root, train=True, download=False, *,
+                 sampling_rate=16000, include=None, silence=True,
+                 transform=None, target_transform=None):
+
+        if download:
+            self.root = safe_path(root)
+            self._download()
```

Comment:
Unfortunately, when you call `super` only at the end of the class definition (which makes totally sense here) the usage of `root` is not safe in all occurences. I would propose to start `__init__` with a line:
```python
self.root = safe_path(root)
```
and afterwards always use `self.root` instead of `root`. Then you also have to call `safe_path` only once.

At the moment I got the following error:
```python
>>> data = SpeechCommands(root='~/tmp/speech_commands_v0.02', download=True)
Extracting /home/audeering.local/hwierstorf/tmp/speech_commands_v0.02.tar.gz
To /home/audeering.local/hwierstorf/tmp/speech_commands_v0.02
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/audeering.local/hwierstorf/git/audeering/audtorch/audtorch/datasets/speech_commands.py", line 80, in __init__
    d = os.listdir(os.path.join(root, speech_cmd))
FileNotFoundError: [Errno 2] No such file or directory: '~/tmp/speech_commands_v0.02/right'
```
even so `~/tmp/speech_commands_v0.02/right exists on my PC.
If I change all the calls to `root` as described above it works.

Reply:
Ah I see, thanks. Do you know why that behaviour occurs? I only ask as I've never seen that before.

Reply:
The problem is that we use `self.root = safe_path(root)` in `AudioDataset`, as normally we always need `safe_path`. But when you call `super` at the very end it will not be available before.
So if you call `super` not as the first part of your data set class, you always should start with `self.root = safe_path(root)`.

This is not ideal, but after a long discussion we were not able to find a better solution up to now.